### PR TITLE
gh-99108: Import SHA2-224 and SHA2-256 from HACL*

### PR DESCRIPTION
**Pull request by msprotz**
Issue #99108 is about replacing hashlib primitives (for the non-OpenSSL case) with verified implementations from HACL*. This is the first PR in the series, and focuses specifically on SHA2-256 and SHA2-224.

This PR imports Hacl_Streaming_SHA2 into the Python tree. This is the HACL* implementation of SHA2, which combines a core implementation of SHA2 along with a layer of buffer management that allows updating the digest with any number of bytes. This supersedes the previous implementation in the tree.

@franziskuskiefer was kind enough to benchmark the changes: in addition to being verified (thus providing significant safety and security improvements), this implementation also provides a sizeable performance boost!

```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
Sha2_256_Streaming            3163 ns      3160 ns       219353     // this PR
LibTomCrypt_Sha2_256          5057 ns      5056 ns       136234     // library used by Python currently
``` 

The changes in this PR are as follows:
- import the subset of HACL* that covers SHA2-256/224 into `Modules/_hacl`
- rewire sha256module.c to use the HACL* implementation

To review the changes, one should focus on `sha256module.c`, and possibly `Hacl_Streaming_SHA2.h` which is the API entry point. The rest of the files are part of HACL* and are cherry-picked straight from the HACL* repository, so as to minimize the amount of hand-edits and facilitate future refreshes of the code, should HACL* land some performance improvements.

The original code comes from https://github.com/project-everest/hacl-star/tree/master/dist/gcc-compatible. The various .h files in `include/`, `krmllib/`, etc. will be shared across (hopefully) future primitives imported from HACL*. I can trim those files to minimize the amount of includes, but then this will make refreshing the code in the future harder (since the trimming will have to be redone for each refresh).

The code was originally authored by @karthikbhargavan and I; @polubelova and @r1km provided a considerable amount of help over the past week, as we were overhauling our implementation to offer a significant performance boost (we want this first PR to be a good one!). Thanks to everyone involved in this team effort.

There are a couple remarks left in the source code, but for now, I think it's better to open up a PR and start discussing. As promised, tagging @alex for this PR. Thanks so much!

* Issue: gh-99108

### CODE REVIEW

**Review by gpshead — COMMENTED** [review]
It'd be useful for this kind of thing to include a script (somewhere under Tools?.. where to put it can be figured out later) that can generate and update the hacl-star vendored code implementation. As I assume we'll not want this to be a one time code drop but periodically update it, at least for every major Python release, as hacl-star evolves.

**Comment by gpshead** [comment]
Please include a benchmark comparing to the OpenSSL based sha2 as well.

File: Modules/Setup.stdlib.in [hunk 1]
```diff
@@ -78,7 +78,7 @@
 # hashing builtins, can be disabled with --without-builtin-hashlib-hashes
 @MODULE__MD5_TRUE@_md5 md5module.c
 @MODULE__SHA1_TRUE@_sha1 sha1module.c
-@MODULE__SHA256_TRUE@_sha256 sha256module.c
+@MODULE__SHA256_TRUE@_sha256 _hacl/sha256module.c _hacl/Hacl_Streaming_SHA2.c -I$(srcdir)/Modules/_hacl/include -I$(srcdir)/Modules/_hacl/internal -I$(srcdir)/Modules/_hacl/krmllib/dist/minimal -DKRML_VERIFIED_UINT128 -D_BSD_SOURCE -D_DEFAULT_SOURCE
```
Comment (tiran) [inline Modules/Setup.stdlib.in]: Includes and defines belong into `PY_STDLIB_MOD` block for sha256module in `configure.ac`.

**Review by tiran — COMMENTED** [review]
Please add all new headers to `MODULE__SHA256_DEPS` in `Makefile.pre.in`.

I would prefer if you keep the `sha256module.c` file in its current place and only replace libtom code with HACL calls. It's easier to review and keeps the history for untouched lines.

**Comment by franziskuskiefer** [comment]
> Please include a benchmark comparing to the OpenSSL based sha2 as well.

Adding to the numbers in the PR description https://github.com/python/cpython/pull/99109#issue-1436683253 we get the following. Note that OpenSSL is compiled without assembly support (`no-asm no-hw`) to get a fair comparison.

```
OpenSSL_Sha2_256           3000 ns         2999 ns       233849
```

**Comment by gpshead** [comment]
> Note that OpenSSL is compiled without assembly support (`no-asm no-hw`) to get a fair comparison.

LOL, thanks, but fair enough. 😄   This does at least answer my broader question about the feasibility of _also_ getting rid of the OpenSSL backed `_hashlib` in the long run: not today.

File: Modules/Setup.stdlib.in (hunk 1 above)
Comment (msprotz) [reply Modules/Setup.stdlib.in]: Thanks for pointing me to the right location... I've fixed this, regenerated the configure script using autoconf-2.69 (to minimize the diff), and also moved sha256module.c (and the clinic-generated file) back into their original locations.

**Comment by msprotz** [comment]
To keep the discussion general and not related to sha256, I responded on the issue regarding performance, ASM vs no-ASM, etc.

I'm still facing build issues, in which the _hacl directory is not being created, and the compiler can't write out the .o file. This seems to be specific to the scenario where the build output directory is outside of the source tree. The error happens early on for Linux, and later on for OSX. If that sounds familiar to anyone (@tiran?) I'd be happy to take pointers on how to fix it. Thanks.

**Comment by tiran** [comment]
Checkout `SRCDIRS` in `configure.ac`.

**Comment by msprotz** [comment]
Thanks to @zooba for helping me fix the Windows build.

@gpshead to answer your earlier comment: while this PR is a "pure C" version, here are our performance numbers for our version (not submitted as part of this PR) that leverages the SHA-NI instruction, when present. This is on a Ryzen 5.

```
EverCrypt_Sha2_256_Streaming        776 ns          776 ns       907329 // universal streaming API for all hashes, but calling SHA-NI when available
OpenSSL_Sha2_256                    524 ns          524 ns      1334337 // with asm optimizations
```

We haven't worked on optimizing this specific codepath, so I'm confident that we could drive the number lower and be almost the same as OpenSSL. This is also on a relatively small sample size, so numbers would probably be almost equal for larger input sizes.

Bottom line is, there are ways to make verified code on-par with OpenSSL... this PR is just a first step.

**Comment by msprotz** [comment]
hi everyone, just a friendly ping on this PR; is there any way I can help the discussion make progress? happy to provide more context, explanations, or perf measurements if any of these would be helpful... thanks!

File: Modules/sha256module.c [hunk 2]
```diff
@@ -67,321 +71,19 @@ _sha256_get_state(PyObject *module)
...
-    digest[23] = (unsigned char) ((sha_info->digest[5]      ) & 0xff);
-    digest[24] = (unsigned char) ((sha_info->digest[6] >> 24) & 0xff);
-    digest[25] = (unsigned char) ((sha_info->digest[6] >> 16) & 0xff);
-    digest[26] = (unsigned char) ((sha_info->digest[6] >>  8) & 0xff);
-    digest[27] = (unsigned char) ((sha_info->digest[6]      ) & 0xff);
-    digest[28] = (unsigned char) ((sha_info->digest[7] >> 24) & 0xff);
-    digest[29] = (unsigned char) ((sha_info->digest[7] >> 16) & 0xff);
-    digest[30] = (unsigned char) ((sha_info->digest[7] >>  8) & 0xff);
-    digest[31] = (unsigned char) ((sha_info->digest[7]      ) & 0xff);
+    // TODO: implement this in HACL
```
Comment (gpshead) [inline Modules/sha256module.c]: If there is an upstream issue on the HACL* repo, drop a link to that here.

File: Modules/sha256module.c [hunk 3]
```diff
@@ -503,7 +200,8 @@ SHA256Type_update(SHAobject *self, PyObject *obj)
 
     GET_BUFFER_VIEW_OR_ERROUT(obj, &buf);
 
-    sha_update(self, buf.buf, buf.len);
+    if (Hacl_Streaming_SHA2_update_256(self->state, buf.buf, buf.len) != 0)
```
Comment (gpshead) [inline Modules/sha256module.c]: buf.len is a 64-bit Py_ssize_t, the HACL* method takes a uint32_t.  (convenient github actions compiler warning)

A loop around update seems necessary. And we need to ensure we've got a bigmem test covering a >4GiB input to update.

Comment (gpshead) [reply Modules/sha256module.c]: Related: We don't want to raise a RuntimeError as our API allows an arbitrary Python length input.

Comment (gpshead) [reply Modules/sha256module.c]: this also applies to the update calls in the constructors below.

**Comment by gpshead** [comment]
> It'd be useful for this kind of thing to include a script (somewhere under Tools?) that can generate and update the hacl-star vendored code implementation. As I assume we'll not want this to be a one time code drop but periodically update it, at least for every major Python release, as hacl-star evolves.

this comment still applies.

File: Modules/sha256module.c (hunk 3 above)
Comment (msprotz) [reply Modules/sha256module.c]: what should be the behavior if the maximum hash length is exceeded? sha2-256 is only defined for inputs of length up to 2^63-1, and undefined otherwise

**Comment by msprotz** [comment]
> this comment still applies.

I'm happy to do that, although I touched up a couple things manually, so it might take a few PRs in HACL to get the generated C code to be exactly like the one I have in this PR.

Would it be ok to add the script as part of a followup PR, or would it be more strategic (meaning, increased likelihood of a PR accepted!) to have it here now along with the initial PR?

File: Modules/sha256module.c (hunk 3 above)
Comment (gpshead) [reply Modules/sha256module.c]: CPython isn't currently supported on any platform where `Py_ssize_t` could be larger than that $(2^{63} - 1)$ so I just wouldn't worry about it.

Have we ever had any code enforce that total sum of all update() method calls length in the past. Does anyone? It's an unreasonably long linear computation.

Comment (msprotz) [reply Modules/sha256module.c]: Regarding handling of data > 4GB: fixed with a small helper, and a suitable test in f4b267ebe24d5b8c33001b11eaab68a8f10e575c that feeds 4GB of zeroes, followed by "hello world" into the hash function.

Regarding checking the return code: I mis-spoke, the max length is 2^61-1. But in any case, even with best-in-class accelerators, it would take over a billion years to have that much data, so I think it's ok to ignore the return value -- this is what I did in the revised version.

File: Modules/sha256module.c (hunk 2 above)
Comment (msprotz) [reply Modules/sha256module.c]: implemented upstream for the purposes of this PR, and fixed in a632c84b6c09e87b51378821f171f1ddce8df6aa

**Comment by msprotz** [comment]
After a few weeks of upstream work, here is a revised PR. Highlights:
- thanks to upstream proofs showing that various internal steps (update, update_blocks, update_last) are identical for SHA2-224 and SHA2-256, the vendored files are about 400 lines smaller now
- almost all of the manual edits that I had performed earlier are now fixed upstream in the verified codebase, meaning that I was able to write a bash script that imports the required files automatically from upstream into the CPython repository
- we now implement the missing state copy function
- I fixed handling of more than 4GB of data, along with a test
- I added a script to automatically refresh the vendored code from upstream.

Regarding that last point, the script:
- cherry-picks the 6 files currently imported into the CPython tree
- manually removes a series of #includes that are not necessary for the specific subset of HACL files we import here
- removes sha2-384 and sha2-512 from the file (I want to keep this initial PR small and easy to review)
- trims the default headers to remove definitions that are in general useful for the HACL* toolchain but not used by SHA2-256 specifically

The last three steps are manual cleanups, and are not strictly necessary: without them, there would be a few more unused definitions and headers. But I want to keep this initial PR neat and tidy, hence the manual elimination of definitions that are generally useful, but not used by SHA2-256 specifically.

**Comment by gvanrossum** [comment]
@gpshead Seems you are the one to see this through to the end. Could you see what's left to do?

**Comment by gpshead** [comment]
Okay, I think this is ready. I pushed a few changes to the PR branch:

* Improvements to refresh.sh.
* A README.md in Modules/_hacl/ to go along with this third party code.
* Used C #define hacks to "namespace" the `Hacl_` prefixed symbols from the .c file so that they can't cause ODR violations if a process winds up linking with other code that linked against its own version of HACL\*. _(we do this on some of our other third party things within CPython)_. It was manually generated and needs manual updating, but I doubt the list will change often. I'd like a test to confirm that no `Hacl_` C symbols exist in `_sha256.so` or `_sha256.pyd`... but that can remain a future todo wishlist item for now.

**Review by gpshead — APPROVED** [review]

File: Modules/sha256module.c [hunk 4]
```diff
@@ -524,12 +228,13 @@ SHA256_get_block_size(PyObject *self, void *closure)
...
+SHA256_get_name(SHAobject *self, void *closure)
 {
-    if (((SHAobject *)self)->digestsize == 32)
-        return PyUnicode_FromStringAndSize("sha256", 6);
-    else
+    if (self->digestsize == 28) {
         return PyUnicode_FromStringAndSize("sha224", 6);
+    } else {
+        return PyUnicode_FromStringAndSize("sha256", 6);
+    }
```
Comment (erlend-aasland) [inline Modules/sha256module.c]:
```suggestion
    }
    return PyUnicode_FromStringAndSize("sha256", 6);
```

**Review by erlend-aasland — COMMENTED** [review]
One small style nit. Also, there's a cast warning on line 127 in Modules/sha256module.c.

Autoconf changes looks good to me.

**Comment by erlend-aasland** [comment]
You'll see it in the diff tab on the PR; it's one of the GitHub windows runners that's complaining, IIRC (I'm on mobile now).

**Comment by msprotz** [comment]
I fixed the final review comments... @erlend-aasland do I need to re-request a review from you, or is there anything I can do? thanks!

**Review by gpshead — APPROVED** [review]
This PR is in good shape!  But on its own - lets call this step "2a" - it'll be a little odd to ship in 3.12 if we don't also have some natural related followups to include as well:

 * [x] 2b) SHA384 & SHA512 support (sha512module.c and related hacl code).
 * [ ] 3) SHA3 support (more generated hacl code).

**I'll go ahead and merge this**, and be responsible for reverting it if we decide we don't wind up just the shorter half of SHA2 using it if the follow-ons aren't ready yet before 3.12-beta1 in May.

Could you go ahead and prepare follow-on PRs?

**Comment by msprotz** [comment]
Yup! I'll respond on the issue so that we don't scatter the overall discussion across individual PRs.
